### PR TITLE
Implement search by name in variables page

### PR DESCRIPTION
**Pull request by cbenz**
Implements #22 and #23 in a row (was easier).

### CODE REVIEW

File: src/components/pages/variables.jsx [hunk 1]
```diff
@@ -32,11 +59,21 @@ const VariablesPage = React.createClass({
     countryPackageVersion: PropTypes.string.isRequired,
     variables: PropTypes.arrayOf(AppPropTypes.variable).isRequired,
   },
+  getInitialState() {
+    return {
+      labelFilter: "",
+      nameFilter: ""
```
Comment (MattiSG) [inline src/components/pages/variables.jsx]: The `filter` suffix evokes a function to me, not a query string. What about `query` or `lookup`?

File: src/components/search-box.jsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    labelFilter: PropTypes.string,
+    nameFilter: PropTypes.string,
+    onLabelChange: PropTypes.func,
+    onNameChange: PropTypes.func,
+  },
+  onLabelInputChange(event) {
+    this.props.onLabelChange(event.target.value)
+  },
+  onNameInputChange(event) {
+    this.props.onNameChange(event.target.value)
```
Comment (MattiSG) [inline src/components/search-box.jsx]: I have limited experience with React, but I thought one of its points was to remove that kind of event forwarding 🤔

File: src/components/search-box.jsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+              className="form-control"
+              id="search-box-name"
+              onChange={this.onNameInputChange}
+              placeholder="Nom"
+              value={this.props.nameFilter}
+            />
+          </div>
+          <div className="form-group">
+            <label htmlFor="search-box-label">
+              Label
```
Comment (MattiSG) [inline src/components/search-box.jsx]: If the UI is in French, this should be translated too.

File: src/components/search-box.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            </label>
+            <input
+              className="form-control"
+              id="search-box-label"
+              onChange={this.onLabelInputChange}
+              placeholder="Label"
+              value={this.props.labelFilter}
```
Comment (MattiSG) [inline src/components/search-box.jsx]: Why separate label and name search? I don't think the choice is obvious for the user.

File: src/components/pages/variables.jsx [hunk 5]
```diff
@@ -1,28 +1,55 @@
...
-    (memo, variable) => over(
-      lensPath(getVariablePath(variable)),
-      node => node
-        ? over(lensPath(["variables"]), append(variable), node)
-        : {variables: [variable]},
-      memo,
-    ),
-    {},
-    variables
+  const matchFilters = variable => (!nameFilter || variable.name.toLowerCase().includes(nameFilter.toLowerCase())) && (
```
Comment (cbenz) [inline src/components/pages/variables.jsx]: OK for that.

Personally I tend to use `function`s when there is a need to declare variables and/or use control flow like `if` or `for` loops, and `=>` functions when there is none of them (this is more a parametrized expression).

But I'm fine with both.

Comment (MattiSG) [inline src/components/pages/variables.jsx]: Well, I think there is a actually a need for `if`, and that it was worked around with `! nameFilter ||`, which is perfectly fine in itself but leads here to a slightly long expression  ;)

File: src/components/pages/variables.jsx [hunk 6]
```diff
@@ -1,28 +1,55 @@
...
+    !labelFilter || variable.label && variable.label.toLowerCase().includes(labelFilter.toLowerCase())
```
Comment (cbenz) [inline src/components/pages/variables.jsx]: I'm not sure to understand your sentence :)

File: src/components/pages/variables.jsx [hunk 7]
```diff
@@ -1,28 +1,55 @@
...
   )
+  function removeEmptyNodes(node) {
+    if (isNil(node.children)) {
+      return node
+    } else {
```
Comment (cbenz) [inline src/components/pages/variables.jsx]: Personally I would have used a 2 ternary operators (well-indented for readability).

Do you like it or prefer a real `if` with early exit?

File: src/components/pages/variables.jsx (hunk 6 above)
Comment (MattiSG) [inline src/components/pages/variables.jsx]: I'm saying [accent-folding](http://alistapart.com/article/accent-folding-for-auto-complete) should be part of #23 :wink:

File: src/components/pages/variables.jsx (hunk 7 above)
Comment (cbenz) [inline src/components/pages/variables.jsx]: Oops, sorry this does not apply in this case due to the variable declaration in the `else`.

File: src/components/pages/variables.jsx [hunk 8]
```diff
@@ -1,28 +1,55 @@
...
+          ? node
+            ? over(lensPath(["variables"]), append(variable), node)
+            : {variables: [variable]}
+          : node,
+        memo,
+      ),
+      {}
+    ),
+    removeEmptyNodes,
+  )(variables)
```
Comment (cbenz) [inline src/components/pages/variables.jsx]: > Why is it so complicated to display a list in React?

This is not related to React.

This function filters the tree leaves, keeping the intermediary non-empty nodes.
It works on data, not the DOM.

Then this data is passed to the DOM in the [`render` method](https://github.com/openfisca/legislation-explorer/pull/35/files/4977ca339fe53affb2f56f32f3c2262d5c8a2f03#diff-d761e4c762b02dbf437c932af71cb9f0R76).

It is expressed using the Ramda library functions (which is like lodash/underscore) and the `pipe` function chains 2 transformations of data. Perhaps this style is complicated but that's a matter of style, and I can rewrite it in a less functional way.

Anyway I'm gonna post a screenshot of the filtered tree to better understand what I did.

File: src/components/search-box.jsx (hunk 2 above)
Comment (cbenz) [inline src/components/search-box.jsx]: Hmm that's an interesting discussion (or a troll ^^) but I'm not gonna explain that here.

I'd just say that:
- React is a view library and does not provide any architecture excepted callbacks for events, which are passed to the parent "component"
- Flux architecture has been proposed and its most widespread implementation is the Redux library
- Redux is not used in Legislation Explorer
- `<troll>React+Redux+Webpack+Babel+Ramda+Typescript+... => Elm language + Elm architecture + Elm toolchain</troll>` and I'm now happy with it, and I see through all the comments you do here that you'd like it!

**Comment by cbenz** [comment]
Here is the rendering:

openfisca variables filter

**Comment by cbenz** [comment]
New rendering:

openfisca variables filter 2

File: src/components/search-box.jsx (hunk 4 above)
Comment (fpagnoux) [inline src/components/search-box.jsx]: 👍

**Comment by cbenz** [comment]
I removed the label filtering, and renamed this PR as a consequence.

**Pull request merged**